### PR TITLE
feat(engine-v2): Add error code for all GraphQL errors

### PR DESCRIPTION
**Pull request by Finistere**
Adding a mandatory error code in engine-v2 errors.
I'm not sure how we should handle custom error codes from hooks though if we want to support that later.

I have renamed the `runtime::error:{GraphqlError, ErrorCode}` to `runtime::error::{PartialGraphqlError, PartialErrorCode}` as they're extended by the engine on top. Not a great name, but avoids having identical names at least.

### CODE REVIEW

File: engine/crates/engine/src/error.rs
Comment (obmarg) [inline engine/crates/engine/src/error.rs:38]: Why do engine-v2 errors live inside engine-v1?

Comment (Finistere) [reply engine/crates/engine/src/error.rs:38]: because the `ErrorCode` was initially created for `PersistedQueryNotFound`. Can create one specific to engine-v2 though if you prefer

Comment (obmarg) [reply engine/crates/engine/src/error.rs:38]: Not going to insist, but seems like it'd make more sense from an organisational perspective.

Comment (obmarg) [reply engine/crates/engine/src/error.rs:38]: (Bearing in mind I have thought about this for all of 30 seconds - feel free to disagree)

File: engine/crates/engine-v2/src/engine.rs
Comment (pimeys) [inline engine/crates/engine-v2/src/engine.rs:288]: If we already call `Response::bad_request`, could this automatically generate a `GraphqlError` with the corresponding error code?

File: engine/crates/engine-v2/src/response/error.rs
Comment (pimeys) [inline engine/crates/engine-v2/src/response/error.rs:17]: This is kind of breaking, because the btree is ordered by the key and this is not. It's probably fine, just hilighting this here.

File: engine/crates/engine-v2/src/response/error.rs
Comment (pimeys) [inline engine/crates/engine-v2/src/response/error.rs:32]: Builder stuff is awesome here.

File: engine/crates/engine-v2/src/response/read/ser.rs
Comment (pimeys) [inline engine/crates/engine-v2/src/response/read/ser.rs:136]: What happens if a hook adds a code to the error extensions. Do we get two codes in the response?

File: engine/crates/engine/src/error.rs
Comment (pimeys) [reply engine/crates/engine/src/error.rs:38]: :+1:

**Review by pimeys — APPROVED** [review]
Approving with caveats to check out before merging.

File: engine/crates/engine-v2/src/response/error.rs
Comment (Finistere) [reply engine/crates/engine-v2/src/response/error.rs:17]: it's serialized as a map, no ordering guarantees in GraphQL, but yeah fair point

File: engine/crates/engine-v2/src/response/read/ser.rs
Comment (Finistere) [reply engine/crates/engine-v2/src/response/read/ser.rs:136]: gets ignored today, I should probably just let it through though

File: engine/crates/engine-v2/src/engine.rs
Comment (Finistere) [reply engine/crates/engine-v2/src/engine.rs:288]: fair point

Comment (Finistere) [reply engine/crates/engine-v2/src/engine.rs:288]: Ok not the same thing, bad request here means it's a request error in the GraphQL-spec sense. So no `data` field. Error could also be validation one or persisted query not found.

Comment (Finistere) [reply engine/crates/engine-v2/src/engine.rs:288]: renamed to `pre_execution_error` which is closer to the spec meaning

**Pull request merged**
